pkg/snapshot/copier/copier.go

```diff
@@ -138,18 +140,54 @@ func (c *Copier) copyBackups() error {
 		destSnapshotsMap[snapshot.SnapName] = snapshot
 	}
 
+	// find snapshots missing in destination
+	var snapshotsToCopy brtypes.SnapList
 	for _, snapshot := range sourceSnapshot {
-		// Copy all the snapshots (if not already copied)
 		if _, ok := destSnapshotsMap[snapshot.SnapName]; !ok {
-			c.logger.Infof("Copying %s snapshot %s...", snapshot.Kind, snapshot.SnapName)
-			if err := c.copySnapshot(snapshot); err != nil {
-				return err
-			}
+			snapshotsToCopy = append(snapshotsToCopy, snapshot)
 		} else {
 			c.logger.Infof("Skipping %s snapshot %s as it already exists", snapshot.Kind, snapshot.SnapName)
 		}
 	}
-	return nil
+
+	if len(snapshotsToCopy) == 0 {
+		return nil
+	}
+
+	// copy all missing snapshots in parallel
+	var (
+		wg        sync.WaitGroup
+		errors    = make(chan error)
+		allErrors error
+	)
+
+	for _, s := range snapshotsToCopy {
+		snapshot := s
+		wg.Add(1)
+
+		go func() {
```

Comment:
It is generally not a good idea to spawn one go routine for each snapshot to copy. Sure, it might be faster, but for cases where there is a large number of snapshots to be copied, the `etcdbrctl copy` command might very easily run out of memory or get cpu-throttled. It's always recommended to use bounded workers for such operations, so that we don't overload the container and still achieve parallelism for copying the snapshots. I would also recommend exposing the number of workers used for copying via a CLI flag like `--num-copiers` so that this can be configured by the user, similar to `--max-fetchers` that is used to configure the maximum number of fetcher goroutines for the restoration operation in etcd-backup-restore.
 https://github.com/gardener/etcd-backup-restore/blob/ba86fb950b139cf6edd8a79c4aff65793af34b8f/pkg/types/restorer.go#L104

Reply:
Agreed, done in a new commit.
I admit, we thought about this originally but dropped the idea because we didn't think about OOM kills. That's a good argument to limit concurrency 👍